app/models/user.rb

```diff
@@ -25,80 +29,165 @@ def to_s
-    !(groups & MANAGER_GROUPS).empty?
+    !(groups & MANAGER_GROUPS).blank?
+  end
+
+  # @return [Boolean] is the user a repository wide viewer
+  def is_viewer
+    !(groups & VIEWER_GROUPS).blank?
   end
 
   # The convention is for boolean is_XYZ methods to be interrogative (end in "?")
   alias_method :is_admin?,   :is_admin
-  alias_method :is_viewer?,  :is_viewer
   alias_method :is_manager?, :is_manager
+  alias_method :is_viewer?,  :is_viewer
 
   def login
     webauth ? webauth.login : sunetid
   end
 
-  ##
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_manage_object?(dor_object)
+    dor_object = parse_object(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
+    manager_access = is_admin || is_manager
+    manager_access ||= dor_object.can_manage_item?(roles(apo_pid))
+    manager_access ||= dor_object.can_manage_content?(roles(apo_pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_item?(roles(pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_content?(roles(pid))
+    manager_access
+  end
+
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_view_object?(dor_object)
+    dor_object = parse_object(dor_object)
+    return true if can_manage_object?(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
+    view_access = is_viewer
+    view_access ||= dor_object.can_view_content?(roles(apo_pid))
+    view_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_view_content?(roles(pid))
+    view_access
+  end
+
+  # Do argo permissions allow viewing anything.
   # @return [Boolean]
   def can_view_something?
     is_admin || is_manager || is_viewer || permitted_apos.length > 0
   end
+
+  private
+
+  # Find a Dor object, whether the input is a DRUID, a PID or a Dor object.
+  # @param druid [String|ActiveFedora::Base]
+  #              DruidTools::Druid is used to parse String, which is
+  #              then input for Dor.find to obtain a Dor object
+  #              Dor objects inherit from ActiveFedora::Base
+  # @raises ArgumentError
+  def parse_object(druid)
```

Comment:
OK, `get_dor_object` it is.  There are two methods that require the same code to determine if the argument is a string (and what kind of string) or a dor object and there are places where either can be given in the specs.  Removing the call to `find_druid` and using the code there directly (despite the duplication, to maintain separation of model and controller and avoid the complications of that method).